### PR TITLE
Update authlib from 0.14.3 to 1.3.1

### PR DESCRIPTION
**Pull request by magopian**
This is needed by [udata#482](https://github.com/datagouv/udata-front/pull/482) to change the SSO login button from MonComptePro to AgentConnect

### CODE REVIEW

**Comment by magopian** [comment]
Oops, I need to update the udata code (and the tests) which are now failing, let me address that before reviewing please

**Review by bolinocroustibat — APPROVED** [review]
Great, thanks, LGTM!

**Comment by maudetes** [comment]
I tried a login process locally with an Oauth2 client (publier.etalab.studio) but it doesn't seem to work anymore (return a 401 on a /api/1/me/ call with an `Authorization` header).

Steps to reproduce:
* git clone https://github.com/etalab/publier.etalab.studio/
* `cp .env.sample .env`
* update .env with
```
VUE_APP_OAUTH_BASE_URL=http://dev.local:7000/fr
VUE_APP_OAUTH_CALLBACK=http://localhost:8080/
VUE_APP_DATAGOUV_API_URL=http://dev.local:7000
```
* npm install & npm run serve
* go to http://localhost:8080/fr/
* start a udata env with `AUTHLIB_INSECURE_TRANSPORT=1 inv serve` (on master VS this branch)
* click on login -> gets redirected to dev.local -> accept -> gets redirected to http://localhost:8080/fr/
* look at the `/api/1/me/` being made

**Comment by magopian** [comment]
Thanks a lot @maudetes for the reproduction steps.
It's my bad, the `save_token` code was completely bogus, and didn't actually save the token.
This should now be fixed, and I updated the tests so we don't run into this issue again in the future ;) (regression tests ftw).

I tested locally and all seems to be working as expected. Could you please do the same on your end and make sure everything is fixed?

File: udata/api/oauth2.py [hunk 1]
```diff
@@ -111,7 +112,7 @@ def check_redirect_uri(self, redirect_uri):
     def check_client_secret(self, client_secret):
         return self.secret == client_secret
 
-    def check_token_endpoint_auth_method(self, method):
+    def check_endpoint_auth_method(self, method, _endpoint):
```
Comment (maudetes) [inline udata/api/oauth2.py:115]: Those additional endpoints and renaming were needed for the new version?

File: udata/api/oauth2.py [hunk 2]
```diff
@@ -161,6 +165,13 @@ def get_expires_at(self):
     def get_client_id(self):
         return str(self.client.id)
 
+    def is_expired(self):
```
Comment (maudetes) [inline udata/api/oauth2.py:168]: I'm slightly confused between `is_expired` VS `is_refresh_token_valid` logic

File: udata/api/oauth2.py [hunk 3]
```diff
@@ -295,7 +306,7 @@ def revoke_token():
 def authorize(*args, **kwargs):
     if request.method == "GET":
         try:
-            grant = oauth.validate_consent_request(end_user=current_user)
+            grant = oauth.get_consent_grant(end_user=current_user)
```
Comment (maudetes) [inline udata/api/oauth2.py:309]: does `get_consent_grant` also `validate`?

File: udata/tests/api/test_auth_api.py [hunk 4]
```diff
@@ -582,7 +584,7 @@ def test_invalid_implicit_grant_token(self, client, oauth):
         )
 
         assert_status(response, 400)
-        assert response.json["error"] == "invalid_grant"
+        assert response.json["error"] == "unsupported_response_type"
```
Comment (maudetes) [inline udata/tests/api/test_auth_api.py:587]: Was this error modification expected?

File: udata/tests/api/test_auth_api.py [hunk 5]
```diff
@@ -606,6 +608,11 @@ def test_refresh_token(self, client, oauth):
         assert200(response)
         assert response.content_type == "application/json"
         assert "access_token" in response.json
+        new_access_token = response.json["access_token"]
+        assert new_access_token != "access-token"  # The access token has been refreshed.
+        tokens = OAuth2Token.objects(client=oauth, user=user)
+        assert len(tokens) == 1  # No new token has been created.
+        assert tokens.first().access_token == new_access_token
```
Comment (maudetes) [inline udata/tests/api/test_auth_api.py:615]: Thank you for this additional test! :pray:

File: udata/api/oauth2.py [hunk 6]
```diff
@@ -324,13 +335,13 @@ def query_client(client_id):
 
 def save_token(token, request):
     scope = token.pop("scope", "")
+    client = request.client
+    user = request.user or client.owner
     if request.grant_type == "refresh_token":
-        credential = request.credential
-        credential.update(scope=scope, **token)
+        old_token = OAuth2Token.objects(client=client, user=user, scope=scope).first()
```
Comment (maudetes) [inline udata/api/oauth2.py:341]: Shouldn't `refresh_token` param be the identifier? Ex: `OAuth2Token.objects(refresh_token=request.refresh_token).first()` or similar?

**Review by maudetes — COMMENTED** [review]
It seems to work as expected! Thank you for the update!

Since I am not familiar with authlib, I have some questions about some modifications I am not sure to understand. You don't need to answer in details if not relevant!

File: udata/api/oauth2.py (hunk 1 above)
Comment (magopian) [reply udata/api/oauth2.py:115]: So what happened is when running the tests after the authlib upgrade, I had a bunch of errors. For each of those errors, I tracked down the initial cause, and search the old and new docs, and try to find how the lib was modified.

This is one of those, but an easy one, as it was right there [in the new docs](https://docs.authlib.org/en/v1.3.1/django/2/authorization-server.html#client)

File: udata/api/oauth2.py (hunk 2 above)
Comment (magopian) [reply udata/api/oauth2.py:168]: `is_expired` seems to be a new method that needs to be implemented in 1.3.1, and is related to the token: is the (access) token still good, or is it expired?

The **refresh** token itself needs to be checked to see if it's still valid. It's valid as long as it's not expired (it's longer lived than the access token), AND has not been revoked.

This allows the revoking mecanism to work properly. Without a short lived access token, even if it's revoked on the auth server (eg AgentConnect's server), the client would still have access to the service (eg datagouv) as long as the access token is alive.

Making the access token short lived forces the client to query for a new one every so often, or to refresh it, at which point the revocation is checked.

I'm not sure if I'm being clear here, but I would be glad to explain this in more details if needed.

File: udata/api/oauth2.py (hunk 3 above)
Comment (magopian) [reply udata/api/oauth2.py:309]: I believe this is a simple function renaming between [0.14.3](https://docs.authlib.org/en/v0.14.3/flask/2/api.html?highlight=validate_consent_request#authlib.integrations.flask_oauth2.AuthorizationServer.validate_consent_request) and [1.3.1](https://docs.authlib.org/en/v1.3.1/specs/rfc6749.html#authlib.oauth2.rfc6749.AuthorizationServer.get_consent_grant), as both their documentation are exactly the same.

I double checked, and indeed in the `get_consent_grant` function, there is a call to `validate_content_request`: https://github.com/lepture/authlib/blob/v1.3.1/authlib/oauth2/rfc6749/authorization_server.py#L217

File: udata/api/oauth2.py (hunk 6 above)
Comment (magopian) [reply udata/api/oauth2.py:341]: Very good point indeed! I modified the code, and the tests to make sure we only update this token, and not any other token.

I've even tested that if a user has two tokens (which shouldn't happen?), only the correct one is refreshed.

File: udata/tests/api/test_auth_api.py (hunk 4 above)
Comment (magopian) [reply udata/tests/api/test_auth_api.py:587]: From what i can see from the source code, the error returned has been changed between [0.14.3](https://github.com/lepture/authlib/blob/v0.14.3/authlib/oauth2/rfc6749/authorization_server.py#L142-L151)  and [1.3.1](https://github.com/lepture/authlib/blob/v1.3.1/authlib/oauth2/rfc6749/authorization_server.py#L198-L207) in the same `get_authorization_grant` function.

File: udata/api/oauth2.py [hunk 7]
```diff
@@ -324,13 +335,15 @@ def query_client(client_id):
...
     scope = token.pop("scope", "")
+    client = request.client
+    user = request.user or client.owner
     if request.grant_type == "refresh_token":
-        credential = request.credential
-        credential.update(scope=scope, **token)
+        old_token = OAuth2Token.objects(
+            refresh_token=request.refresh_token.refresh_token, client=client, user=user, scope=scope
+        ).first()
+        old_token.update(**token)
```
Comment (maudetes) [inline udata/api/oauth2.py:344]: Reading it again, should we return an HTTP error check if old_token is None?

File: udata/api/oauth2.py (hunk 2 above)
Comment (maudetes) [reply udata/api/oauth2.py:168]: Yes I think I get the idea now, thank you! :pray:

**Review by maudetes — APPROVED** [review]
Thank you for all the detailed and sourced explanations!

I've added a suggestion, but else I think we're good to go!

File: udata/api/oauth2.py (hunk 7 above)
Comment (magopian) [reply udata/api/oauth2.py:344]: Slightly unrelated: I went back through the documentation, and saw [this](https://docs.authlib.org/en/latest/flask/2/grants.html#refresh-token-grant):
> By default, RefreshTokenGrant will not issue a refresh_token in the token response. Developers can change this behavior with:
>
> class RefreshTokenGrant(grants.RefreshTokenGrant):
>     INCLUDE_NEW_REFRESH_TOKEN = True

And I believe this is a good idea, so the refresh token is refreshed each and every time the token is refreshed.

More related to your point: I haven't seen anywhere in the code of the lib that would use the return of this function, and I feel reluctant to possibly change the behaviour.

What eases my mind is that we should never reach a state where we're trying to save a token that has been previously validated and accepted through `RefreshTokenGrant.authenticate_refresh_token`.

If my explanation has convinced you, let's land this, and if not, I'd be happy to dig further in this issue ;)

**Pull request merged**